src/main/java/org/opensearch/ad/feature/SearchFeatureDao.java

```diff
@@ -190,67 +233,211 @@ public void getHighestCountEntities(AnomalyDetector detector, long startTime, lo
+             *                               "missing_bucket": false,
+             *                               "order": "asc"
+             *                           }
+             *                       }
+             *                   }]
+             *               },
+             *               "aggregations": {
+             *                   "bucketSort": {
+             *                       "bucket_sort": {
+             *                           "sort": [{
+             *                               "_count": {
+             *                                   "order": "desc"
+             *                               }
+             *                           }],
+             *                           "from": 0,
+             *                           "size": 5,
+             *                           "gap_policy": "SKIP"
+             *                       }
+             *                   }
+             *               }
+             *           }
+             *       }
+             *   }
+             *
+             */
+            bucketAggs = AggregationBuilders
+                .composite(
+                    AGG_NAME_TOP,
+                    detector.getCategoryField().stream().map(f -> new TermsValuesSourceBuilder(f).field(f)).collect(Collectors.toList())
+                )
+                .size(pageSize)
+                .subAggregation(
+                    PipelineAggregatorBuilders
+                        .bucketSort("bucketSort", Arrays.asList(new FieldSortBuilder("_count").order(SortOrder.DESC)))
+                        .size(maxEntitiesForPreview)
+                );
         }
 
         SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder()
             .query(boolQueryBuilder)
-            .aggregation(termsAgg)
+            .aggregation(bucketAggs)
             .trackTotalHits(false)
             .size(0);
         SearchRequest searchRequest = new SearchRequest().indices(detector.getIndices().toArray(new String[0])).source(searchSourceBuilder);
-        ActionListener<SearchResponse> termsListener = ActionListener.wrap(response -> {
-            Aggregations aggs = response.getAggregations();
-            if (aggs == null) {
-                listener.onResponse(Collections.emptyList());
-                return;
-            }
-            List<Entity> results = aggs
-                .asList()
-                .stream()
-                .filter(agg -> AGG_NAME_TERM.equals(agg.getName()))
-                .flatMap(agg -> ((Terms) agg).getBuckets().stream())
-                .map(bucket -> bucket.getKeyAsString())
-                .collect(Collectors.toList())
-                .stream()
-                .map(entityValue -> parseCategoricalField(entityValue, detector))
-                .collect(Collectors.toList());
-            listener.onResponse(results);
-        }, listener::onFailure);
-        client.search(searchRequest, termsListener);
+        client
```

Comment:
For another PR (i.e. non-blocker): I realize this is not being changed in this PR, but why is the node client used here and not a separate threadpool? Isn't this call expensive and it will be run frequently?

Reply:
Sending a search request will land in Opensearch's search threadpool no matter what.  We cannot run it in other threadpool.  If preparing the search request is expensive, we can wrap that logic in the AD threadpool.  Here it is not the case.